src/Day16/Solution.hs

```diff
@@ -0,0 +1,92 @@
+module Day16.Solution
+  ( Document (..),
+    ValidDocument (..),
+    asValidDocument,
+    extractTicket,
+    parseDocument,
+    part1,
+    part2,
+    resolvePossibilities,
+    ticketScanningErrors,
+  )
+where
+
+import Advent.Utils (fromRightOrShowError)
+import Data.Bifunctor (Bifunctor (second))
+import qualified Data.IntMap.Strict as IntMap
+import qualified Data.IntSet as IntSet
+import Data.List (isPrefixOf, sortOn)
+import qualified Data.Map.Strict as Map
+import Day16.Parser (Document (..), Ticket, parseDocument)
+
+part1 :: String -> String
+part1 = show . sum . ticketScanningErrors . fromRightOrShowError . parseDocument
+
+part2 :: String -> String
+part2 = show . product . Map.filterWithKey (const . isPrefixOf "departure") . extractTicket . asValidDocument . fromRightOrShowError . parseDocument
+
+newtype ValidDocument = ValidDocument Document deriving (Show, Eq)
```

Comment:
ValidDocument is something like [`NonEmpty`](https://hackage.haskell.org/package/base-4.14.1.0/docs/Data-List-NonEmpty.html#t:NonEmpty).  If we can guarantee certain features about the quality of the data within the data structure, then we can safely run an algorithm like `extractTicket`.